packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -453,31 +478,37 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
      */
     public reSubmit(metadata: Record<string, unknown> | undefined) {
         assert(!!metadata, 0x38b /* Resubmitted ops must have metadata */);
-        const { blobId, localId }: { blobId?: string; localId?: string; } = metadata;
+        const { localId, blobId }: { localId?: string; blobId?: string } = metadata;
+        assert(localId !== undefined, "local ID not available on reSubmit");
         if (!blobId) {
-            assert(!!localId, 0x38c /* Submitted BlobAttach ops must have a blobId or localId */);
             // We submitted this op while offline. The blob should have been uploaded by now.
             const pendingEntry = this.pendingBlobs.get(localId);
             assert(pendingEntry?.status === PendingBlobStatus.OfflinePendingOp &&
                 !!pendingEntry?.storageId, 0x38d /* blob must be uploaded before resubmitting BlobAttach op */);
-            return this.sendBlobAttachOp(pendingEntry.storageId, localId);
+            return this.sendBlobAttachOp(localId, pendingEntry.storageId);
         }
-        return this.sendBlobAttachOp(blobId, localId);
+        return this.sendBlobAttachOp(localId, blobId);
     }
 
     public processBlobAttachOp(message: ISequencedDocumentMessage, local: boolean) {
-        assert(message?.metadata?.blobId, 0x12a /* "Missing blob id on metadata" */);
-        if (message.metadata.localId !== undefined) {
-            this.redirectTable.set(message.metadata.localId, message.metadata.blobId);
+        const blobId = message.metadata?.blobId;
+        assert(blobId !== undefined, 0x12a /* "Missing blob id on metadata" */);
+
+        // Set up a mapping from local ID to storage ID. This is crucial since without this the blob cannot be
+        // requested from the server.
+        if (message.metadata?.localId !== undefined) {
+            this.setRedirection(message.metadata.localId, message.metadata.blobId);
         }
         // set identity (id -> id) entry
-        this.redirectTable.set(message.metadata.blobId, message.metadata.blobId);
+        this.setRedirection(blobId, message.metadata.blobId);
 
         if (local) {
-            if (message.metadata.localId === undefined) {
-                // Since there is no local ID, we know this op was submitted while online.
-                const waitingBlobs = this.opsInFlight.get(message.metadata.blobId);
-                assert(!!waitingBlobs, 0x38e /* local online BlobAttach op with no pending blob */);
+            assert(message.metadata?.localId !== undefined, "local ID not present in blob attach message");
+            const waitingBlobs = this.opsInFlight.get(blobId);
+            if (waitingBlobs !== undefined) {
```

Comment:
It's ok to have undefined waitingBlobs here?

Reply:
Yes. This will happen in scenarios where there are multiple ops in flight for the same storage id, When the first op is received, all pending ops will be resolved and the opsInFlight entry will be deleted. The next set of ops will simply be no-op.